### PR TITLE
Android | reduce minSdkVersion to 19

### PR DESCRIPTION
**Pull request by tiholic**
Any sdkVersion with appropriate source and target compatibility configuration will work just fine

### CODE REVIEW

**Review by QuintinWillison — CHANGES_REQUESTED** [review]
I note the integration testing workflow is only testing levels 24, 26 and 28. Can you add some more to the matrix as part of this PR?

**Comment by kavalerov** [comment]
I suggest we add levels down to 19 (as this is the minimum supported version of our `ably-java` SDK)

File: README.md [hunk 1]
```diff
@@ -25,6 +25,23 @@ or visit our
 [Support and Help](https://www.ably.io/support)
 site to discuss privately.
 
+## Supported Platforms
+
+#### iOS
+
+iOS9 or newer
```
Comment (QuintinWillison) [inline README.md:32]:
```suggestion
iOS 9 or newer.
```

File: README.md [hunk 2]
```diff
@@ -25,6 +25,23 @@ or visit our
...
+
+#### Android
+
+SDKVersion 19 or newer.
```
Comment (QuintinWillison) [inline README.md:36]: API Level 19 (Android 4.4, KitKat) or newer.

File: README.md [hunk 3]
```diff
@@ -25,6 +25,23 @@ or visit our
...
+
+_This project uses Java8 language features and uses [Desugaring](https://developer.android.com/studio/write/java8-support#library-desugaring)
```
Comment (QuintinWillison) [inline README.md:38]:
```suggestion
_This project uses Java 8 language features, utilising [Desugaring](https://developer.android.com/studio/write/java8-support#library-desugaring)
```

File: README.md [hunk 4]
```diff
@@ -25,6 +25,23 @@ or visit our
...
+to support lower versions of android (< SDKVersion 24)_
```
Comment (QuintinWillison) [inline README.md:39]:
```suggestion
to support lower versions of the Android runtime (i.e. API Levels prior to 24)_
```

File: README.md [hunk 5]
```diff
@@ -25,6 +25,23 @@ or visit our
...
+
+If your project needs support for lower SDK Version, Android Gradle Plugin 4.0.0+ must be used.
```
Comment (QuintinWillison) [inline README.md:41]: Lower than what? Can you make this more explicit please.

File: README.md [hunk 6]
```diff
@@ -25,6 +25,23 @@ or visit our
...
+You might also need to upgrade [gradle distribution](https://developer.android.com/studio/releases/gradle-plugin#updating-plugin) accordingly.
+
+
```
Comment (QuintinWillison) [inline README.md:44]: Please avoid multiple new lines. One is always enough. 😁

File: android/build.gradle [hunk 7]
```diff
@@ -23,21 +23,26 @@ apply plugin: 'com.android.library'
...
-    compileSdkVersion 28
+    compileSdkVersion 30
 
     // Require Java language level 8 so we can use Method References (used with Lambdas)
     compileOptions {
-        sourceCompatibility 1.8
-        targetCompatibility 1.8
+        sourceCompatibility JavaVersion.VERSION_1_8
+        targetCompatibility JavaVersion.VERSION_1_8
+        // Flag to enable Java 8 language APIs while supporting to older SDK Versions
```
Comment (QuintinWillison) [inline android/build.gradle:36]: Please insert a blank line above this comment to make it clearer that it lives with the line below.

File: android/build.gradle [hunk 8]
```diff
@@ -23,21 +23,26 @@ apply plugin: 'com.android.library'
...
+        coreLibraryDesugaringEnabled true
     }
 
     defaultConfig {
-        // Require Android 7.0 (Nougat) runtime so we can use Lambda expressions
-        minSdkVersion 24
-
+        minSdkVersion 19
+        // Required when setting minSdkVersion to 20 or lower so we can use
```
Comment (QuintinWillison) [inline android/build.gradle:42]: Please insert a blank line above this comment block to make it clearer that it lives with the line below.

**Review by QuintinWillison — CHANGES_REQUESTED** [review]
Also, I notice you've explicitly only upgraded the Gradle wrappers to 6.5, but the latest is [6.7.1](https://github.com/gradle/gradle/releases/tag/v6.7.1). Any reason you couldn't upgrade all the way up?

File: android/build.gradle (hunk 8 above)
Comment (tiholic) [reply android/build.gradle:44]: Addressed: a23b04ef17039c4a68d05220ea76366e3e0e974e

File: android/build.gradle (hunk 7 above)
Comment (tiholic) [reply android/build.gradle:37]: Addressed: a23b04ef17039c4a68d05220ea76366e3e0e974e

File: README.md (hunk 6 above)
Comment (tiholic) [reply README.md]: Addressed: a23b04ef17039c4a68d05220ea76366e3e0e974e

File: README.md (hunk 5 above)
Comment (tiholic) [reply README.md]: Addressed: a23b04ef17039c4a68d05220ea76366e3e0e974e

File: README.md (hunk 4 above)
Comment (tiholic) [reply README.md]: Addressed: a23b04ef17039c4a68d05220ea76366e3e0e974e

File: README.md (hunk 3 above)
Comment (tiholic) [reply README.md]: Addressed: a23b04ef17039c4a68d05220ea76366e3e0e974e

File: README.md (hunk 2 above)
Comment (tiholic) [reply README.md]: Addressed: a23b04ef17039c4a68d05220ea76366e3e0e974e

File: README.md (hunk 1 above)
Comment (tiholic) [reply README.md]: Addressed: a23b04ef17039c4a68d05220ea76366e3e0e974e

**Comment by tiholic** [comment]
@QuintinWillison thanks for pointing out. I've upgraded gradle wrapper to 6.7.1

**Review by QuintinWillison — APPROVED** [review]

**Pull request merged**